### PR TITLE
Add upgrade logic to kubernetes anywhere

### PR DESCRIPTION
**Pull request by jessicaochen**
Issue #431 

Adds structure for upgrades and fills in the upgrade implementation for kubeadm on gce.

Design doc: https://docs.google.com/a/google.com/document/d/1PoDeqD8qu1KbAd-fb7xOVB1G7YTyD6VLLlhaVB3QeA4/edit?usp=sharing

### CODE REVIEW

File: phase1/gce/configure-vm-kubeadm.sh [hunk 1]
```diff
@@ -3,6 +3,8 @@
 TOKEN=$(get_metadata "k8s-kubeadm-token")
 KUBEADM_VERSION=$(get_metadata "k8s-kubeadm-version")
 KUBERNETES_VERSION=$(get_metadata "k8s-kubernetes-version")
+KUBEADM_DIR=/etc/kubeadm
+KUBEADM_INIT_PARAM_FILE=$KUBEADM_DIR/kubadm_init_params.txt
```
Comment (luxas) [inline phase1/gce/configure-vm-kubeadm.sh]: typo kubeadm

File: phase2/Kconfig [hunk 2]
```diff
@@ -39,6 +39,14 @@ config phase2.kubeadm.version
 
 	  Valid options are "stable" or a Google Cloud Storage link to a build.
 
+config phase2.kubeadm.master_upgrade.method
+	string "kubeadm master upgrade method"
+	default "upgrade"
+	help
+	  The kubeadm method of master upgrade to use.
+
+	  Valid options are (upgrade,init).
```
Comment (luxas) [inline phase2/Kconfig]: Didn't we change this now to `kubeadm-upgrade` and `kubeadm-init` and not scope the upgrade functionality under kubeadm?

File: phase2/kubeadm/do [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+set -o errexit
+set -o pipefail
+set -o nounset
+set -x
+
+cd "${BASH_SOURCE%/*}"
+
+CONFIG_FILE="${CONFIG_JSON_FILE:-../../.config.json}"
+KUBEADM_INIT_PARAM_FILE=/etc/kubeadm/kubadm_init_params.txt
```
Comment (luxas) [inline phase2/kubeadm/do]: may need to update typo here as well

File: phase2/kubeadm/do [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  wait_nodes_ready
+
+  INIT_PARAM=$(execute_host ${MASTER} "sudo cat ${KUBEADM_INIT_PARAM_FILE}")
+  case "${UPGRADE_METHOD}" in
+    "init")
+      execute_host ${MASTER} "sudo kubeadm init --skip-preflight-checks --kubernetes-version ${KUBERNETES_VERSION} ${INIT_PARAM}"
+      ;;
+    "upgrade")
+      execute_host ${MASTER} "sudo kubeadm config upload from-flags ${INIT_PARAM}"
+      execute_host ${MASTER} "sudo kubeadm upgrade apply ${KUBERNETES_VERSION} -y -f"
```
Comment (luxas) [inline phase2/kubeadm/do]: nit: no need to do `-f` with `-y`, but why are you setting `-f`?

File: phase2/kubeadm/do [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  echo host is ${HOST} version is ${KUBEADM_VERSION}
+  if [[ "${KUBEADM_VERSION}" == stable ]]; then
+    execute_host ${HOST} "echo '  deb http://apt.kubernetes.io/ kubernetes-xenial main' | sudo tee /etc/apt/sources.list.d/kubernetes.list"
+    execute_host ${HOST} "sudo apt-get update"
+    execute_host ${HOST} "sudo apt-get install -y kubelet kubeadm kubectl kubernetes-cni"
+  elif [[ "${KUBEADM_VERSION}" == "gs://"* ]]; then
+    TMPDIR=/tmp/k8s-debs
+    execute_host ${HOST} "sudo mkdir ${TMPDIR}"
+    execute_host ${HOST} "sudo gsutil rsync "${KUBEADM_VERSION}" ${TMPDIR}"
+    execute_host ${HOST} "sudo dpkg -i ${TMPDIR}/{kubelet,kubeadm,kubectl,kubernetes-cni}.deb || echo Ignoring expected dpkg failure"
```
Comment (luxas) [inline phase2/kubeadm/do]: As discussed, kubeadm and kubelet should be decoupled

File: phase2/Kconfig (hunk 2 above)
Comment (jessicaochen) [reply phase2/Kconfig]: kubeadm-init is not formally supported. However, we do need this here so we can use it for e2e testing. I can hide it it the comments to make it less confusing what is supported and what is not supported.

File: phase1/gce/configure-vm-kubeadm.sh (hunk 1 above)
Comment (jessicaochen) [reply phase1/gce/configure-vm-kubeadm.sh]: done

File: phase2/kubeadm/do (hunk 3 above)
Comment (jessicaochen) [reply phase2/kubeadm/do]: done

File: phase2/kubeadm/do (hunk 5 above)
Comment (jessicaochen) [reply phase2/kubeadm/do]: Done

File: phase2/kubeadm/do (hunk 4 above)
Comment (jessicaochen) [reply phase2/kubeadm/do]: Removed with changes for the other comments.

**Comment by jessicaochen** [comment]
@luxas ready for next review round

File: phase2/Kconfig (hunk 2 above)
Comment (pipejakob) [reply phase2/Kconfig]: After looking at the previous code, @luxas's comment, and the new code, I think you might have misinterpreted his request. He can correct me, but I think he was just saying that we should call these "kubeadm-upgrade" and "kubeadm-init" because these methods are specific to kubeadm. Upgrading isn't available for ignition yet (or maybe ever?), but if it were, it might have other methods like "ignition-rocket" and "ignition-swapinplace." Adding the prefix just makes it very clear what the method is used by.

I'm also not sure what you mean by "kubeadm-init is not formally supported." Do you mean within the context of kubernetes-anywhere? Because at least for kubeadm, upgrading clusters via `kubeadm init` is definitely supported, which is the reason we want it e2e tested in the first place.

File: phase2/Kconfig [hunk 6]
```diff
@@ -39,6 +39,14 @@ config phase2.kubeadm.version
 
 	  Valid options are "stable" or a Google Cloud Storage link to a build.
 
+config phase2.kubeadm.master_upgrade.method
+	string "kubeadm master upgrade method"
+	default "upgrade"
```
[the remark this answers is not in the document]
Comment (pipejakob) [reply phase2/Kconfig]: Ignore that comment entirely -- I thought I had deleted it from the review when I answered my own questions.

File: phase2/Kconfig [hunk 7]
```diff
@@ -39,6 +39,14 @@ config phase2.kubeadm.version
...
+	help
+	  The kubeadm method of master upgrade to use.
+
+	  Valid options are upgrade.
```
Comment (pipejakob) [inline phase2/Kconfig]: I would definitely list all valid options, instead of hiding any. If we do have different levels of support or recommendations for the options to use, we can document those in the help text, but in a project whose primary goals are to make everything easy to read and understand, it seems wrong to deliberately omit options.

File: phase2/kubeadm/do [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+set -x
+
+cd "${BASH_SOURCE%/*}"
+
+CONFIG_FILE="${CONFIG_JSON_FILE:-../../.config.json}"
+KUBEADM_INIT_PARAM_FILE=/etc/kubeadm/kubeadm_init_params.txt
+
+upgrade_master() {
+  KUBERNETES_VERSION=$(jq -r '.phase2.kubernetes_version' ${CONFIG_FILE})
+  UPGRADE_METHOD=$(jq -r '.phase2.kubeadm.master_upgrade.method' ${CONFIG_FILE})
```
Comment (pipejakob) [inline phase2/kubeadm/do]: Per @luxas' suggestion, instead of having `.phase2.kubeadm.*` for upgrade options, they do seem generic enough to move to just `.phase2` (e.g. `.phase2.upgrade_strategy`).

Also, is there a need for having the upgrade method scoped to the master here? I'm having trouble imagining a scenario where a normal user would want to use different upgrade strategies on the master versus nodes, so I'm wondering if there are specific test scenarios you're trying to enable.

File: phase2/kubeadm/do [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+  MASTER=$(do_phase1 master-host)
+
+  update_host_kubeadm ${MASTER}
+  wait_nodes_ready
+
+  INIT_PARAM=$(execute_host ${MASTER} "sudo cat ${KUBEADM_INIT_PARAM_FILE}")
+  case "${UPGRADE_METHOD}" in
+    "init")
+      execute_host ${MASTER} "sudo kubeadm init --skip-preflight-checks --kubernetes-version ${KUBERNETES_VERSION} ${INIT_PARAM}"
```
Comment (pipejakob) [inline phase2/kubeadm/do]: Have you tested how this handles cases like filenames with spaces or escaped characters? Those are always tricky to get right.

File: phase2/kubeadm/do [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  exit 1
+}
+
+update_host_kubeadm() {
+  HOST=$1 
+  KUBEADM_VERSION=$(jq -r '.phase2.kubeadm.version' ${CONFIG_FILE})
+  KUBELET_VERSION=$(jq -r '.phase2.kubelet_version' ${CONFIG_FILE})
+  if [[ "${KUBEADM_VERSION}" == stable ]]; then
+      # TODO(jessicaochen): Figure out how to translate "stable" to the corresponding GS link so we can
+      # update kubadm without disturbing the kubelet. Installing kubeadm via package will modify the kubelet configuration.
```
Comment (pipejakob) [inline phase2/kubeadm/do]: s/kubadm/kubeadm/

File: phase2/kubeadm/do [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+      sleep 5
+    fi
+  done
+  echo "Gave up waiting for nodes to become ready." >&2
+  exit 1
+}
+
+update_host_kubeadm() {
+  HOST=$1 
+  KUBEADM_VERSION=$(jq -r '.phase2.kubeadm.version' ${CONFIG_FILE})
```
Comment (pipejakob) [inline phase2/kubeadm/do]: I know this was done in another PR, but the inconsistency of these two is pretty striking when seen like this. It would be great to get them both using the dot or underscore.

`.phase2.kubeadm.version`
`.phase2.kubelet_version`

File: phase2/kubeadm/do [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+  KUBELET_VERSION=$(jq -r '.phase2.kubelet_version' ${CONFIG_FILE})
+  if [[ "${KUBEADM_VERSION}" == stable ]]; then
+      # TODO(jessicaochen): Figure out how to translate "stable" to the corresponding GS link so we can
+      # update kubadm without disturbing the kubelet. Installing kubeadm via package will modify the kubelet configuration.
+      echo "Kubeadm version of 'stable' is not yet supported for upgrades. Please specify a Google Cloud Storage Link."
+      exit 1
+  elif [[ "${KUBEADM_VERSION}" == "gs://"* ]]; then
+    TMPDIR=/tmp/k8s-debs
+    execute_host ${HOST} "sudo mkdir ${TMPDIR}"
```
Comment (pipejakob) [inline phase2/kubeadm/do]: Of these four execute_host commands, only the third (copying into /usr/bin) actually looks like it needs to be run under sudo, yet all of them are.

**Comment by pipejakob** [comment]
/lgtm

No blockers from me, just minor suggestions that can also be addressed as follow-up PRs later if need be. Otherwise, awesome job!

File: phase2/kubeadm/do [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+  echo "Gave up waiting for nodes to become ready." >&2
+  exit 1
+}
+
+update_host_kubeadm() {
+  HOST=$1 
+  KUBEADM_VERSION=$(jq -r '.phase2.kubeadm.version' ${CONFIG_FILE})
+  KUBELET_VERSION=$(jq -r '.phase2.kubelet_version' ${CONFIG_FILE})
+  if [[ "${KUBEADM_VERSION}" == stable ]]; then
+      # TODO(jessicaochen): Figure out how to translate "stable" to the corresponding GS link so we can
```
Comment (luxas) [inline phase2/kubeadm/do]: `curl -sSL dl.k8s.io/release/$(curl -sSL dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubeadm`
please fix this

File: phase1/gce/do [hunk 14]
```diff
@@ -88,6 +88,27 @@ destroy() {
...
+  MASTER=$(jq -r '.phase1.cluster_name' ${CONFIG_FILE})-master
+  SSHUSER=$(jq -r '.phase1.ssh_user | values' ${CONFIG_FILE})
+
+  if [[ -n "${SSHUSER}" ]]; then
+    SSHUSER="${SSHUSER}@"
+  elif [[ -n "${USER:-}" ]]; then
+    SSHUSER="${USER}@"
+  fi
+
+  gcloud compute ssh --project "$PROJECT" --zone "$ZONE" "$SSHUSER$MASTER" --command "$2" >&1
```
Comment (luxas) [inline phase1/gce/do]: nit: `"${SSHUSER}${MASTER}"`

File: phase2/kubeadm/do [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+  KUBERNETES_VERSION=$(jq -r '.phase2.kubernetes_version' ${CONFIG_FILE})
+  UPGRADE_METHOD=$(jq -r '.phase2.kubeadm.master_upgrade.method' ${CONFIG_FILE})
+  MASTER=$(do_phase1 master-host)
+
+  update_host_kubeadm ${MASTER}
+  wait_nodes_ready
+
+  INIT_PARAM=$(execute_host ${MASTER} "sudo cat ${KUBEADM_INIT_PARAM_FILE}")
+  case "${UPGRADE_METHOD}" in
+    "init")
```
Comment (luxas) [inline phase2/kubeadm/do]: `kubeadm-init`

File: phase2/kubeadm/do [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+      execute_host ${MASTER} "sudo kubeadm init --skip-preflight-checks --kubernetes-version ${KUBERNETES_VERSION} ${INIT_PARAM}"
+      ;;
+    "upgrade")
```
Comment (luxas) [inline phase2/kubeadm/do]: `kubeadm-upgrade`
See: https://github.com/kubernetes/test-infra/pull/4495

File: phase2/kubeadm/do [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+      execute_host ${MASTER} "sudo kubeadm config upload from-flags ${INIT_PARAM}"
+      execute_host ${MASTER} "sudo kubeadm upgrade apply ${KUBERNETES_VERSION} -y -f"
+      ;;
+  esac
+}
+
+wait_nodes_ready() {
```
Comment (luxas) [inline phase2/kubeadm/do]: nit: don't we have this code somewhere else already?

File: phase2/kubeadm/do [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+  HOST=$1 
+  KUBEADM_VERSION=$(jq -r '.phase2.kubeadm.version' ${CONFIG_FILE})
+  KUBELET_VERSION=$(jq -r '.phase2.kubelet_version' ${CONFIG_FILE})
+  if [[ "${KUBEADM_VERSION}" == stable ]]; then
+      # TODO(jessicaochen): Figure out how to translate "stable" to the corresponding GS link so we can
+      # update kubadm without disturbing the kubelet. Installing kubeadm via package will modify the kubelet configuration.
+      echo "Kubeadm version of 'stable' is not yet supported for upgrades. Please specify a Google Cloud Storage Link."
+      exit 1
+  elif [[ "${KUBEADM_VERSION}" == "gs://"* ]]; then
+    TMPDIR=/tmp/k8s-debs
```
Comment (luxas) [inline phase2/kubeadm/do]: use `mktemp -d /tmp/k8s-debsXXXXXX)` consistently when creating temp dirs in order to not have conflicts (which would be a mess to debug)

**Comment by luxas** [comment]
Also needs a rebase and https://github.com/kubernetes/kubernetes-anywhere/pull/452 to actually work

File: phase2/kubeadm/do (hunk 8 above)
Comment (jessicaochen) [reply phase2/kubeadm/do]: This config is specifically how do you want kubeadm to upgrade the master/control plane. It has no relevance to how the node is updated. It is also not an overall upgrade strategy.

I can see your point if it was an name for an overall upgrade strategy or if it was a setting that would make sense for other phase2 providers (eg. should I drain the node before update).

File: phase1/gce/configure-vm-kubeadm.sh [hunk 19]
```diff
@@ -55,7 +58,11 @@ case "${ROLE}" in
     if [[ -n "$KUBERNETES_VERSION" ]]; then
       OPTS="--kubernetes-version $KUBERNETES_VERSION"
     fi
-    kubeadm init --token "${TOKEN}" --apiserver-bind-port 443 --skip-preflight-checks --apiserver-advertise-address "$(get_metadata "k8s-advertise-addresses")" $OPTS
+    ADVERTISE_ADDRESS=$(get_metadata "k8s-advertise-addresses")
+    PARAMS="--token ${TOKEN} --apiserver-bind-port 443 --apiserver-advertise-address ${ADVERTISE_ADDRESS}"
+    kubeadm init $PARAMS --skip-preflight-checks $OPTS
+    mkdir $KUBEADM_DIR
+    echo "${PARAMS}" | tee $KUBEADM_INIT_PARAM_FILE
```
[the remark this answers is not in the document]
Comment (jessicaochen) [reply phase1/gce/configure-vm-kubeadm.sh]: I remember there was a reason I ended up going with tee. I will get back on this later assuming it is not a blocking issue from your standpoint.

File: phase2/kubeadm/do (hunk 9 above)
Comment (jessicaochen) [reply phase2/kubeadm/do]: I have not. I have the init param filename and path hardcoded so I do control if the filename has spaces or escaped characters. I will revisit this if we ever introduce the opportunity for spaces or escaped characters to the filename.

File: phase2/kubeadm/do (hunk 11 above)
Comment (jessicaochen) [reply phase2/kubeadm/do]: Will need to think about this and the right approach going forward. I followed the pattern of kubernetes_version since the kubelet is a required part of a cluster just like the control plane is.

File: phase2/kubeadm/do (hunk 12 above)
Comment (jessicaochen) [reply phase2/kubeadm/do]: You are probably right. I will revisit later.

File: phase2/kubeadm/do (hunk 15 above)
Comment (jessicaochen) [reply phase2/kubeadm/do]: I think there has been a miscommunication. I am fine with the test using that in its own param. For the actual k8s anywhere config, the extra kubeadm is redundant. 

ie.
.phase2.kubeadm.master_upgrade.method=kubeadm-init
.phase2.kubeadm.master_upgrade.method=init

File: phase2/kubeadm/do (hunk 11 above)
Comment (luxas) [reply phase2/kubeadm/do]: Fine for now

File: phase2/kubeadm/do (hunk 16 above)
Comment (jessicaochen) [reply phase2/kubeadm/do]: I think there has been a miscommunication. I am fine with the test using that in its own param. For the actual k8s anywhere config, the extra kubeadm is redundant. 

ie.
.phase2.kubeadm.master_upgrade.method=kubeadm-upgrade
.phase2.kubeadm.master_upgrade.method=init

File: phase2/kubeadm/do (hunk 10 above)
Comment (jessicaochen) [reply phase2/kubeadm/do]: Done.

File: phase1/gce/configure-vm-kubeadm.sh [hunk 20]
```diff
@@ -4,6 +4,10 @@ TOKEN=$(get_metadata "k8s-kubeadm-token")
 KUBEADM_VERSION=$(get_metadata "k8s-kubeadm-version")
 KUBERNETES_VERSION=$(get_metadata "k8s-kubernetes-version")
 KUBELET_VERSION=$(get_metadata "k8s-kubelet-version")
+KUBEADM_DIR=/etc/kubeadm
+KUBEADM_INIT_PARAM_FILE=$KUBEADM_DIR/kubeadm_init_params.txt
+KUBELET_VERSION=$(get_metadata "k8s-kubelet-version")
```
Comment (luxas) [inline phase1/gce/configure-vm-kubeadm.sh]: rebase conflict?

File: phase1/gce/configure-vm-kubeadm.sh [hunk 21]
```diff
@@ -51,15 +55,19 @@ fi
 
 case "${ROLE}" in
   "master")
-    OPTS=''
+    ADVERTISE_ADDRESS=$(get_metadata "k8s-advertise-addresses")
+    PARAMS="--token ${TOKEN} --apiserver-bind-port 443 --apiserver-advertise-address ${ADVERTISE_ADDRESS}"
```
Comment (luxas) [inline phase1/gce/configure-vm-kubeadm.sh]: nit: shouldn't have both PARAMS and OPTS, choose one

File: phase2/kubeadm/do (hunk 4 above)
Comment (luxas) [reply phase2/kubeadm/do]: I don't understand.

File: phase2/kubeadm/do (hunk 16 above)
Comment (luxas) [reply phase2/kubeadm/do]: Well, as long as you've tested this end to end and what merged in https://github.com/kubernetes/test-infra/pull/4495 works, I'm ok with this

File: phase2/kubeadm/do (hunk 13 above)
Comment (jessicaochen) [reply phase2/kubeadm/do]: Done.

File: phase2/kubeadm/do (hunk 18 above)
Comment (jessicaochen) [reply phase2/kubeadm/do]: Need to address this across the board. Created Issue to track. https://github.com/kubernetes/kubernetes-anywhere/issues/453

File: phase1/gce/do (hunk 14 above)
Comment (jessicaochen) [reply phase1/gce/do]: Done.

File: phase1/gce/configure-vm-kubeadm.sh (hunk 21 above)
Comment (jessicaochen) [reply phase1/gce/configure-vm-kubeadm.sh]: OPTS are the options that do not get saved and passed to kubadm init later. PARAMS are saved and passed on. I do need to separate them. Is there perhaps a better name choice or something that would make it more clear.

File: phase1/gce/configure-vm-kubeadm.sh (hunk 20 above)
Comment (jessicaochen) [reply phase1/gce/configure-vm-kubeadm.sh]: Bad merge on my end. Let me remove the extra kubelet version.

File: phase2/kubeadm/do (hunk 4 above)
Comment (jessicaochen) [reply phase2/kubeadm/do]: Found a way to properly reply. -f is needed to install non-stable versions eg. beta 1.8

Comment (luxas) [reply phase2/kubeadm/do]: @jessicaochen You should ideally use `--allow-experimental-upgrades` for that purpose
